core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1539,6 +1539,13 @@ abstract class RDD[T: ClassTag](
     // NOTE: we use a global lock here due to complexities downstream with ensuring
     // children RDD partitions point to the correct parent partitions. In the future
     // we should revisit this consideration.
+    if (doCheckpointCalled) {
+      throw new SparkException("Because job has been executed on this RDD, checkpoint won't work")
+    }
```

Comment:
It is not an error condition to call checkpoint() on an RDD which has already been checkpoint'ed : for example in complex workflow, different modules might call checkpoint as an invariant - and the RDD might already have been persisted/checkpointed/computed.
Since isCheckpointed returns true only after checkpoint succeeds (and not if checkpoint() was invoked), there is no way for these methods to prevent an exception.